include/zephyr/shell/shell.h

```diff
@@ -1241,6 +1290,143 @@ int shell_set_root_cmd(const char *cmd);
+ *
+ * As the shell receives characters from its underlying transport layer, it puts
+ * them into a pipe, which the caller must drain out by calling
+ * shell_bypass_receive(). If the caller does not drain characters fast enough,
+ * they will be dropped. If the caller's application encounters this situation,
+ * it can increase the size of the receive pipe using configuration parameter
+ * CONFIG_SHELL_BYPASS_USER_RX_PIPE_SIZE.
+ *
+ * Note that this is a syscall that uses the shell pointer directly rather than
+ * the shell_bypass pointer. It is valid to call this function at any time and
+ * from any thread.
+ *
+ * The count is reset to 0 each time bypass mode is entered.
+ *
+ * @param[in] sh	Pointer to the shell instance.
+ *
+ * @retval		The number of dropped characters.
+ */
+__syscall size_t shell_bypass_get_rx_dropped(const struct shell *sh);
+
+/**
+ * @brief Poll for characters received by the shell's transport layer while in
+ * user bypass mode.
+ *
+ * The caller must obtain a valid struct shell_bypass by entering user bypass
+ * mode with shell_bypass_enter(). Only the thread that invoked the enter
+ * method may call shell_bypass_receive().
+ *
+ * This function uses a pipe whose size and alignment are specified by compile-
+ * time configuration variables SHELL_BYPASS_USER_RX_PIPE_SIZE and
+ * SHELL_BYPASS_USER_RX_PIPE_ALIGN, respectively. A pipe buffer whose size is
+ * too small may result in received characters being dropped.
+ *
+ * @param[in] shb		Pointer to shell user bypass structure.
+ * @param[in] data		Address to place the received data.
+ * @param[in] len		Requested number of bytes to receive.
+ * @param[in] timeout		Waiting period to wait for the data to be
+ *				received.
+ *
+ * @retval number		Number of bytes successfully read.
+ * @retval -EPERM		Invoked from invalid thread.
+ * @retval -EINVAL		Invalid parameters supplied.
+ * @retval -EAGAIN		Timeout expired and no data was received.
+ */
+int shell_bypass_receive(struct shell_bypass *shb, void *data, size_t len,
+			 k_timeout_t timeout);
+
+/**
+ * @brief Send bytes to the shell's underlying transport layer.
+ *
+ * The caller must obtain a valid struct shell_bypass by entering user bypass
+ * mode with shell_bypass_enter(). Only the thread that invoked the enter
+ * method may call shell_bypass_send().
+ *
+ * This function uses a pipe whose size and alignment are specified by compile-
+ * time configuration variables SHELL_BYPASS_USER_TX_PIPE_SIZE and
+ * SHELL_BYPASS_USER_TX_PIPE_ALIGN, respectively. A pipe buffer whose size is
+ * too small may result in increased CPU overhead on large data sends due to
+ * frequent thread context switching.
+ *
+ * @param[in] shb		Pointer to shell user bypass structure.
+ * @param[in] data		Address of data to send
+ * @param[in] len		Requested number of bytes to send.
+ * @param[in] timeout		Waiting period to wait for the data to be sent.
+ *
+ * @retval number		Number of bytes sucessfully sent.
```

Comment:
Should this say "Number of bytes passed to underlying transport layer"? Regarding the description of `shell_bypass_exit()` the data is not necessarily successfully sent is it?